weblab/fitting/views.py

```diff
@@ -197,3 +204,150 @@ def get(self, request, *args, **kwargs):
+    and (where relevant) their versions.
+    """
+    permission_required = 'fitting.run_fits'
+    form_class = FittingResultCreateForm
+
+    template_name = 'fitting/fittingresult_create_form.html'
+
+    def get_initial(self):
+        initial = super().get_initial()
+        if 'model' in self.request.GET:
+            initial['model'] = get_object_or_404(ModelEntity, pk=self.request.GET['model'])
+        elif 'protocol' in self.request.GET:
+            initial['protocol'] = get_object_or_404(ProtocolEntity, pk=self.request.GET['protocol'])
+        elif 'fittingspec' in self.request.GET:
+            initial['fittingspec'] = get_object_or_404(FittingSpec, pk=self.request.GET['fittingspec'])
+        elif 'dataset' in self.request.GET:
+            initial['dataset'] = get_object_or_404(Dataset, pk=self.request.GET['dataset'])
+
+        return initial
+
+    def post(self, request, *args, **kwargs):
+        form = self.get_form()
+        if form.is_valid():
+            return self.form_valid(form)
+        else:
+            return self.form_invalid(form)
+
+    def form_valid(self, form):
+        self.runnable, is_new = submit_fitting(
+            form.cleaned_data['model_version'],
+            form.cleaned_data['protocol_version'],
+            form.cleaned_data['fittingspec_version'],
+            form.cleaned_data['dataset'],
+            self.request.user,
+            True,
+        )
+
+        queued = self.runnable.status == FittingResultVersion.STATUS_QUEUED
+
+        if is_new:
+            if queued:
+                messages.info(self.request, "Fitting experiment submitted to the queue.")
+            else:
+                messages.error(self.request, "Fitting experiment could not be run: " + self.runnable.return_text)
+        else:
+            messages.info(self.request, "Fitting experiment was already run.")
+
+        return super().form_valid(form)
+
+    def get_success_url(self):
+        if hasattr(self, 'runnable'):
+            return reverse('fitting:result:version', args=[self.runnable.fittingresult.pk, self.runnable.pk])
+
+
+class FittingResultFilterJsonView(LoginRequiredMixin, PermissionRequiredMixin, View):
+    """
+    JSON view of valid fitting result input values based on those already selected
+
+    For example, if a model id is specified (as a GET param), only versions of that model
+    (which are visible to the user) will be included in the results. Otherwise all visible
+    models and versions will be in the results (which are simply a list of database
+    IDs of the relevant objects)
+
+    Connections between protocols, fitting specs and datasets are also enforced.
+    """
+    permission_required = 'fitting.run_fits'
+
+    def get(self, request, *args, **kwargs):
+        options = {}
+
+        models = ModelEntity.objects.visible_to_user(request.user)
+        model_versions = CachedModelVersion.objects.visible_to_user(request.user)
```

Comment:
Given we're doing the visibility query to get versions, would it be more efficient to make `models` a query based on `model_versions` etc? Something like `models = ModelEntity.objects.filter(cachedmodel__versions__in=model_versions)`?